[tcplp] add support for TCP Fast Open (without cookie management)

samkumar opened the pull request:
This pull request adds support for TCP Fast Open, without cookie management.

To add support for this, I looked at the FreeBSD codebase and brought in some code from FreeBSD 12.0 that implements TCP Fast Open --- the version of FreeBSD that TCPlp is based on did not fully support TCP Fast Open.

Here are some packet captures, collected using two nrf52840-DK boards and a packet sniffer, and the CLI commands I ran to give a sense of how TCP Fast Open works: [tfo-captures.zip](https://github.com/openthread/openthread/files/11713563/tfo-captures.zip)

Normally, a part of TFO is cookie management --- the server generates a cookie and includes it in the initial handshake, and client is expected to present this cookie on future handshakes. This part is not yet implemented, and I changed the logic from FreeBSD to allow data to be exchanged in the TFO handshake even if the client does not present a cookie. If we implement this functionality for TFO later, it is probably worth departing from FreeBSD's data structures and policies for maintaining cookie state in favor of something that is simpler and more memory-efficient. I'm happy to open a discussion about this.

This has been in progress for a long time --- I had to pause work on this as I was interviewing for academic jobs last semester --- but I'm glad to finally be making a pull request.

abtink commented on include/openthread/tcp.h:
```diff
@@ -395,17 +395,23 @@ otError otTcpBind(otTcpEndpoint *aEndpoint, const otSockAddr *aSockName);
 {
-    OT_TCP_CONNECT_NO_FAST_OPEN = 1 << 0,
+    OT_TCP_CONNECT_DISABLE_FAST_OPEN = 1 << 0,
+    OT_TCP_CONNECT_ENABLE_FAST_OPEN  = 1 << 1,
```
What does it mean if we have both `1<<0` and `1<<1` bits set in a given flag?

How about using a single bit-flag to indicate "fast-open" status? The bit-flag would be set to enable fast-open, and cleared to disable it.

samkumar replied:
OK --- I reverted this to the old API with the single `OT_TCP_CONNECT_NO_FAST_OPEN` flag.

Whether TCP Fast Open should be enabled by default or not is a point worth additional discussion. The argument for having TFO enabled by default is that it encourages programmers to use the feature and provides performance closer to a UDP-based solution. @StuartCheshire prefers the existing API for this reason. I'm happy to change the API if there's a consensus on having TFO disabled by default but let's discuss that carefully first.

abtink commented on src/cli/cli_tcp.cpp:
```diff
@@ -908,6 +944,25 @@ void TcpExample::CompleteBenchmark(void)
 }
 
 #if OPENTHREAD_CONFIG_TLS_ENABLE
+void TcpExample::PrepareTLSHandshake(void)
```
We generally treat abbreviation as a name and use `CamelCase` for them (e.g. `MleRouter`, `TcpExample` :)).
```suggestion
void TcpExample::PrepareTlsHandshake(void)
```

samkumar replied:
done

abtink commented on src/cli/cli_tcp.cpp:
```diff
@@ -908,6 +944,25 @@ void TcpExample::CompleteBenchmark(void)
+    mTlsHandshakeComplete = false;
+}
+
 bool TcpExample::ContinueTLSHandshake(void)
```
I know it is existing method, but may be we can rename this as well? `ContinueTlsHandshake()`?

samkumar replied:
done

abtink commented on tests/scripts/expect/cli-tcp-tfo-tls.exp:
```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/expect -f
+#
+#  Copyright (c) 2022, The OpenThread Authors.
```
```suggestion
#  Copyright (c) 2023, The OpenThread Authors.
```

samkumar replied:
done

abtink commented on tests/scripts/expect/cli-tcp-tfo.exp:
```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/expect -f
+#
+#  Copyright (c) 2022, The OpenThread Authors.
```
```suggestion
#  Copyright (c) 2023, The OpenThread Authors.
```

samkumar replied:
done

abtink reviewed (commented):
Thanks @samkumar. Looks good overall. 👍 
Couple of questions/suggestions below:

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -168,17 +168,33 @@ Error Tcp::Endpoint::Bind(const SockAddr &aSockName)
+        sin6p.sin6_port = HostSwap16(aSockName.mPort);
+        error           = BsdErrorToOtError(tcp6_usr_connect(&tp, &sin6p));
+    }
+    else if (fastOpenFlags == 0)
```
Can remove `fastOpenFalgs == 0` check ? and use `else` directly here? 

----

Actually may be more clear if we remove the `fastOpenFalgs` variable and do a direct check in `if`?
```c++
if (aFlags & OT_TCP_CONNECT_NO_FAST_OPEN) 
{
     // ... (code when do not have fast open)
}
else 
{
     // ... fast-open mode
}
```

samkumar replied:
Thanks --- I just pushed an update addressing this.

abtink reviewed (approved):
LGTM. Thanks @samkumar 
One small nit suggestion below.

samkumar commented:
@abtink @jwhui I addressed your comments and rebased on `main`.

I also realized that, when I updated the CLI tool to add an argument for TCP Fast Open, I forgot to edit the README. So I updated the latest commit to include that and pushed that change as well.

jwhui reviewed (approved):
Thanks! 👍🏼
